File: src/main/java/org/blacksun/graph/algorithms/BellmanFordAlgorithm.java
package org.blacksun.graph.algorithms;

import org.blacksun.graph.channel.Channel;
import org.blacksun.graph.node.GraphNode;
import org.blacksun.utils.Pair;
import org.jetbrains.annotations.NotNull;

import java.util.Comparator;
import java.util.HashMap;
import java.util.List;

public class BellmanFordAlgorithm implements PathFindingAlgorithm {
    private final List<GraphNode> nodes;
    private final List<Channel> links;
    private HashMap<Pair<GraphNode, Integer>, Integer> map;
    private HashMap<Pair<GraphNode, Integer>, GraphNode> pathMap;
    private GraphNode last;

    public BellmanFordAlgorithm(@NotNull List<GraphNode> nodes, @NotNull List<Channel> links) {
        this.nodes = nodes;
        this.links = links;
        last = null;
    }

    private void computePaths(@NotNull GraphNode from) {
        if (from.equals(last)) // already computed, return cached
            return;
        last = from;
        map = new HashMap<>();
        pathMap = new HashMap<>();
        int size = nodes.size();
        for (GraphNode node : nodes) {
            for (int i = 0; i < size; ++i) {
                map.put(new Pair<>(node, i), INFINITY);
            }
        }
        map.put(new Pair<>(from, 0), 0);
        for (int i = 1; i < size; ++i) {
            for (Channel channel : links) {
                Pair<GraphNode, Integer> vi = new Pair<>(channel.getToNode(), i);
                GraphNode fromNode = channel.getFromNode();
                Pair<GraphNode, Integer> ui = new Pair<>(fromNode, i - 1);
                Integer oldWeight = map.get(ui);
                if (oldWeight.equals(INFINITY))
                    continue;
                int newWeight = oldWeight + channel.getWeight();
                if (map.get(vi) > newWeight) {
                    map.put(vi, newWeight);
                    pathMap.put(vi, fromNode);
                }
            }
        }
    }

    @Override
    public boolean isAccessible(@NotNull GraphNode from, @NotNull GraphNode to) {
        if (!nodes.contains(from) || !nodes.contains(to))
            return false;
        computePaths(from);
        return map.keySet()
                .stream()
                .anyMatch(pair -> pair.getFirst().equals(to) && !map.get(pair).equals(INFINITY));
    }

    @Override
    public int getMinDistance(@NotNull GraphNode from, @NotNull GraphNode to) {
        if (!nodes.contains(from) || !nodes.contains(to))
            return INFINITY;
        computePaths(from);
        return map.get(minKey(to));
    }

    private Pair<GraphNode, Integer> minKey(GraphNode to) {
        return map.keySet()
                .stream()
                .filter(pair -> pair.getFirst().equals(to))
                .min(Comparator.comparingInt(map::get))
                .orElseThrow(() -> new RuntimeException("Empty graph detected"));
    }

    @Override
    public HashMap<GraphNode, Integer> getDistances(@NotNull GraphNode node) {
        HashMap<GraphNode, Integer> result = new HashMap<>();
        computePaths(node);
        map.forEach((pair, weight) -> {
            GraphNode key = pair.getFirst();
            Integer old = result.get(key);
            if (old == null || old > weight) {
                result.put(key, weight);
            }
        });
        return result;
    }

    @Override
    public GraphPath getPath(@NotNull GraphNode from, @NotNull GraphNode to) {
        GraphPath path = new GraphPath();
        if (!nodes.contains(from) || !nodes.contains(to))
            return path;
        computePaths(from);
        int size = minKey(to).getSecond();
        GraphNode[] nodes = new GraphNode[size + 1];
        for (int j = size; j > 0; --j) {
            nodes[j] = to;
            to = pathMap.get(new Pair<>(to, j));
            if (to == null)
                return path;
        }
        nodes[0] = from;
        for (int i = 0; i < size; ++i) {
            path.add(nodes[i].getConnection(nodes[i + 1]));
        }
        return path;
    }
}

File: src/main/java/org/blacksun/graph/algorithms/BellmanFordAlgorithm.kt
package org.blacksun.graph.algorithms

import org.blacksun.graph.channel.Channel
import org.blacksun.graph.node.GraphNode
import org.blacksun.utils.Pair
import java.util.*

class BellmanFordAlgorithm(private val nodes: List<GraphNode>, private val links: List<Channel>) : PathFindingAlgorithm {
    private var map: HashMap<Pair<GraphNode, Int>, Int> = HashMap()
    private var pathMap: HashMap<Pair<GraphNode, Int>, GraphNode>? = null
    private var last: GraphNode? = null

    init {
        last = null
    }

    private fun computePaths(from: GraphNode) {
        if (from == last)
        // already computed, return cached
            return
        last = from
        map.clear()
        pathMap = HashMap()
        val size = nodes.size
        for (node in nodes) {
            for (i in 0 until size) {
                map[Pair(node, i)] = PathFindingAlgorithm.INFINITY
            }
        }
        map[Pair(from, 0)] = 0
        for (i in 1 until size) {
            for (channel in links) {
                val vi = Pair(channel.toNode, i)
                val fromNode = channel.fromNode
                val ui = Pair(fromNode, i - 1)
                val oldWeight = map[ui]
                if (oldWeight == PathFindingAlgorithm.INFINITY)
                    continue
                val newWeight = oldWeight!! + channel.weight
                if (map[vi]!! > newWeight) {
                    map[vi] = newWeight
                    pathMap!![vi] = fromNode
                }
            }
        }
    }

    override fun isAccessible(from: GraphNode, to: GraphNode): Boolean {
        if (!nodes.contains(from) || !nodes.contains(to))
            return false
        computePaths(from)
        return map.keys
                .stream()
                .anyMatch { pair -> pair.first == to && map[pair] != PathFindingAlgorithm.INFINITY }
    }

    override fun getMinDistance(from: GraphNode, to: GraphNode): Int {
        if (!nodes.contains(from) || !nodes.contains(to))
            return PathFindingAlgorithm.INFINITY
        computePaths(from)
        return map[minKey(to)]!!
    }

    private fun minKey(to: GraphNode): Pair<GraphNode, Int> {
        return map.keys
                .stream()
                .filter { pair -> pair.first == to }
                .min(Comparator.comparingInt { map[it]!! })
                .orElseThrow { RuntimeException("Empty graph detected") }
    }

    override fun getDistances(node: GraphNode): HashMap<GraphNode, Int> {
        val result = HashMap<GraphNode, Int>()
        computePaths(node)
        map.forEach { pair, weight ->
            val key = pair.first
            val old = result[key]
            if (old == null || old > weight) {
                result[key] = weight
            }
        }
        return result
    }

    override fun getPath(from: GraphNode, to: GraphNode): GraphPath {
        var to = to
        val path = GraphPath()
        if (!nodes.contains(from) || !nodes.contains(to))
            return path
        computePaths(from)
        val size = minKey(to).second
        val nodes = arrayOfNulls<GraphNode>(size + 1)
        for (j in size downTo 1) {
            nodes[j] = to
            to = pathMap!![Pair(to, j)] ?: return path
        }
        nodes[0] = from
        for (i in 0 until size) {
            path.add(nodes[i]!!.getConnection(nodes[i + 1]!!))
        }
        return path
    }
}
